htlcswitch/switch.go

```diff
@@ -0,0 +1,599 @@
+
+	Updates chan *lnrpc.CloseStatusUpdate
+	Err     chan error
+}
+
+// pendingPayment represents the payment which made by user and waits for
+// updates to be sent in channels about has it been rejected or processed
+// successfully.
+type pendingPayment struct {
+	err      chan error
+	preimage chan [sha256.Size]byte
+}
+
+// Switch is a central messaging bus for all incoming/outgoing htlc's.
+// The goal of the switch is forward the incoming/outgoing htlc messages from
+// one channel to another, and also propagate the settle/fail htlc messages
+// back to original requester. In order to better understand the whole view:
+//	alice			   bob				   carol
+//	server - <-connection-> - server - - <-connection-> - - - server
+//	 |		   	  |				    |
+//   alice htlc			bob htlc		       carol htlc
+//     switch			switch			         switch
+//	|			 |       \			   |
+//	|			 |        \			   |
+// alice       <-------->  first bob    second bob   <---------> carol
+// channel link	    	  channel link   channel link		channel link
+type Switch struct {
+	started  int32
+	shutdown int32
+	wg       sync.WaitGroup
+	quit     chan bool
+	mutex    sync.RWMutex
+
+	// pendingPayments is correspondence of payments and its hashes,
+	// which is used to save the payments which made by user.
+	pendingPayments map[[sha256.Size]byte]*pendingPayment
+
+	// circuits is an structure which is used to forward the settle/fail
+	// htlc updates back to the add htlc initiator.
+	circuits *circuitMap
+
+	// links is a map of channel output and channel link which manages
+	// this channel.
+	links map[wire.OutPoint]ChannelLink
+
+	// commands is used to propagate the commands to the handler goroutine.
+	commands chan interface{}
+}
+
+// New creates the new instance of htlc switch.
+func New() *Switch {
+	return &Switch{
+		circuits:        newCircuitMap(),
+		links:           make(map[wire.OutPoint]ChannelLink),
+		pendingPayments: make(map[[sha256.Size]byte]*pendingPayment),
+		commands:        make(chan interface{}),
+		quit:            make(chan bool, 1),
+	}
+}
+
+// SendUpdate is used by other subsystems which aren't belong to htlcswicth
+// package in order to send the payment.
+// TODO(andrew.shvv) lnwire messages should be hidden from upper levels.
+func (s *Switch) SendUpdate(dest routing.HopID, htlc *lnwire.UpdateAddHTLC) (
+	chan [sha256.Size]byte, chan error) {
+
+	// Create payment and add to the map of payment in order later to be
+	// able to retrieve it and return response to the user.
+	payment := &pendingPayment{
+		err:      make(chan error, 1),
+		preimage: make(chan [sha256.Size]byte, 1),
+	}
+	s.pendingPayments[htlc.PaymentHash] = payment
+
+	// Generate and send new update packet, if error will be received
+	// on this stage it means that packet didn't left boundaries of our
+	// system and something wrong happened.
+	var source wire.OutPoint
+	if err := s.Forward(newUpdatePacket(source, dest, htlc)); err != nil {
+		payment.err <- err
```

Comment:
In the case of an error, this'll block callers indefinitely, as the `preimage` channel isn't sent upon.